### PR TITLE
Blog post for "Making a GitHub Action"

### PR DESCRIPTION
**Pull request by daisieh**

### CODE REVIEW

**Review by ljdursi — CHANGES_REQUESTED** [review]
Cool, great!  Can you add the necessary Jekyll top matter to the post, like say from the previous post:

```
---
layout: blogpost
title: Federated Learning on CanDIG
summary: An introduction to federated learning and CanDIG-specific implementation considerations.
image: /img/posts/federated-learning-candig/federated-averaging.png
author: Rishabh Sambare
date: 2021-12-22
---
```

Otherwise there's no formatting:  https://7710bca5.candig-github-io.pages.dev/posts/making-github-action/

File: _posts/2022-01-20-making-github-action.md [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+
+GitHub Actions are one possible solution to maintaining this consistency. While GitHub Actions are often used for maintaining development workflows, including integration, deployment, and testing, there's no reason that we can't use Actions to automatically trigger any other required scriptable activity. For example, we can use an Action to maintain the connection between a repository's xlsx files and their csv counterparts.
+
+I created an action to do just that: https://github.com/CanDIG/xlsx2csv-action. Fundamentally, an Action is an executable, self-contained container that runs some code. The container will be executed by GitHub in an enclosing container, allowing each action's container to access the enclosing environment.The Action uses a manifest file to describe how its container will interact with the hosting container. With a GitHub Action, we can guarantee that we are executing in an environment with all of the required packages installed. Environment variables are defined to access inputs and the common workspace.
```
Comment (ljdursi) [inline _posts/2022-01-20-making-github-action.md]: Can you make `https://github.com/CanDIG/xlsx2csv-action` this a markdown link?  Also, could you put a space after the period  in "enclosing environment.The"?

File: _posts/2022-01-20-making-github-action.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+So, to begin creating this action, we write a simple Python script that performs the conversion on an inputted file. This Python file, at its most basic, uses the xlsx2csv package to convert xlsx files to csvs. We can do a little more than that: the script can walk through the repository's files (temporarily stored in the action's workspace) and convert all xlsx files it finds.
+
+```
```
Comment (ljdursi) [inline _posts/2022-01-20-making-github-action.md]: ~There's one more setting to fix before we turn on syntax highlighting, so we don't take advantage of it yet,~ but could you add the language tags to the fenced code blocks: eg, \`\`\`python?

(Also can the indentation be reduced for this code snippet?)

edited to add - actually scratch that, I've added the syntax highlighting bit, so once the blocks are decorated with the language it should be highlighted, as with Rishabh's article

File: _posts/2022-01-20-making-github-action.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        for file in files:
+            file_match = re.match("(.+)\.xlsx$", file)
+            if file_match is not None:
+                output_dir = f"{os.environ.get('GITHUB_WORKSPACE')}/{file_match.group(1)}"
+                Xlsx2csv(file, outputencoding="utf-8").convert(output_dir, sheetid=0)
+```
+
+We can then wrap this in the standard shell script for GitHub Actions, entrypoint.sh:
+
+```
```
Comment (ljdursi) [inline _posts/2022-01-20-making-github-action.md]: Can you mark this as bash e.g. \`\`\`bash?

File: _posts/2022-01-20-making-github-action.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+#!/bin/sh
+
+python /convert.py $*
+```
+
+And then execute that script from a Dockerfile:
+
+```
```
Comment (ljdursi) [inline _posts/2022-01-20-making-github-action.md]: \`\`\`dockerfile

File: _posts/2022-01-20-making-github-action.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+COPY . /
+
+RUN pip install -r requirements.txt
+
+ENTRYPOINT ["/entrypoint.sh"]
+```
+
+The nice thing about using Docker is that you can easily create a Docker environment on your local machine that replicates the environment that the Action will execute in. It'll look something like this docker-compose.yml:
+
+```
```
Comment (ljdursi) [inline _posts/2022-01-20-making-github-action.md]: I think it's \`\`\`yaml although maybe there's a specific docker-compose highlight

File: _posts/2022-01-20-making-github-action.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+  action:
+    build:
+      context: .
+    environment:
+      GITHUB_WORKSPACE: "."
+```
+
+This is very convenient for testing: we can basically create and execute our container locally and be sure that everything runs:
+
+```
```
Comment (ljdursi) [inline _posts/2022-01-20-making-github-action.md]: \`\`\`bash ; yaml for the ones below

File: _posts/2022-01-20-making-github-action.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+I created an action to do just that: https://github.com/CanDIG/xlsx2csv-action. Fundamentally, an Action is an executable, self-contained container that runs some code. The container will be executed by GitHub in an enclosing container, allowing each action's container to access the enclosing environment.The Action uses a manifest file to describe how its container will interact with the hosting container. With a GitHub Action, we can guarantee that we are executing in an environment with all of the required packages installed. Environment variables are defined to access inputs and the common workspace.
+
+So, to begin creating this action, we write a simple Python script that performs the conversion on an inputted file. This Python file, at its most basic, uses the xlsx2csv package to convert xlsx files to csvs. We can do a little more than that: the script can walk through the repository's files (temporarily stored in the action's workspace) and convert all xlsx files it finds.
```
Comment (ljdursi) [inline _posts/2022-01-20-making-github-action.md:19]: Can you link to the package page (maybe https://github.com/dilshod/xlsx2csv, or the pip page?)

File: _posts/2022-01-20-making-github-action.md [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+      - name: Commit changes
+        uses: EndBug/add-and-commit@v4
+        with:
+          author_name: xlsx2csv bot
+          message: "convert xlsx files to csvs"
+          add: "*"
+          ref: ${{env.GITHUB_REF}}
+```
+
+Once this workflow metadata file is pushed to GitHub, this workflow will be executed every time the triggering event occurs. Confusingly, you can see the workflow runs in the Actions tab of your repository, e.g. https://github.com/CanDIG/xlsx2csv-action/actions.
```
Comment (ljdursi) [inline _posts/2022-01-20-making-github-action.md:125]: Can you make `https://github.com/CanDIG/xlsx2csv-action/actions` a markdown link?

**Review by ljdursi — CHANGES_REQUESTED** [review]
This is very clear, thanks!  No real text suggestions, just some markdown stuff.  When that's done, feel free to merge!

File: _posts/2022-01-20-making-github-action.md (hunk 1 above)
Comment (daisieh) [reply _posts/2022-01-20-making-github-action.md]: done

**Pull request merged**